README.md
# Apache log-parser
## About ##
![Alt text](/img/exec.png)


PHP-cli app that implements a parser which reads the Apache access log and saves its content to a database.

## Synopsis ##

    php parser.php [ -f ] file [ -d ] database [ -t ] table
## Description ##
This command line can be used to execute a parser which reads the Apache access log, parses it and saves it to a database

- It parses a standard apache access.log file as reference
- It saves the input into a Mysql Table
- LogFormat: `%h %l %u %t % r %>s %b common`.

Speed sample:
- Computations    :8297 ms
- System calls    :5500 ms
- Total lines     :86506 lines

**Log sample line:**

        127.0.0.1 - frank [10/Oct/2000:13:55:36 -0700] "GET /apache_pb.gif HTTP/1.0" 200 2326
## Options ##


- -f: (optional) Contains the url to the file `default value: /data/access_log`
- -d: (optional) Contains the Database name (created if not exist) `default value: apache_log`
- -t: (optional) Contains the table name (created if not exist) `default value: log`


## Features ##

- I included an access log sample within the folder for test purpose.
- There is no need to create the database nor the table as this scripts creates  one for each with default values. However these params can be changed directly inside the config file `/Conf/config.php` or passed directly as parameters in command line (see example of usage).
- This script also provides a time analysis for the exec time.
## Limits ##
Please change the user and  password for database directly within the config file if it is not the default root user and empty password.
## Example ##
This is an Example of usage of this script:

- ``php parser.php -h`` : This command displays the help for this script.
![Alt text](/img/help.png)
- ``php parser.php`` : This command executes the script with dafault values.
![Alt text](/img/exec.png)
- ``php parser.php -f "path_to_file" -d "sampleDB" -t "sapmleTable"``:
This command parses file path_to_file and saves data inside sapmleTable within sampleDB


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.2] - 2019-04-30
### Fixed
- Optimizing save speed in the database up to 60%.
- Created table is now called log-(timestamp)

### Added
- This CHANGELOG file to hopefully serve as an evolving example of a standardized open source project CHANGELOG.
- Added more logging text to display evolution of process
-

## [0.0.1] - 2017-12-22
### Added
- First commit

test.php
<?php
/**
 * Created by PhpStorm.
 * User: kharr
 * Date: 22/02/2017
 * Time: 23:15
 */
$options = getopt("f:hp:");
var_dump($options);


parser.php
<?php
$rustart = getrusage();

include_once "LogParser.php";
include_once "Controller/ApacheLogController.php";
include_once "Conf/config.php";

try {
    passthru('clear');
    echo $header;
    echo "Script start...\n";
    if ($options = getopt("f::d::t::h::")) {

        if (isset($options['h'])) {
            echo $help;
            die();
        }
        if (isset($options['f']))
            $log_file = $options['f'];
        if (isset($options['d']))
            $db = $options['d'];
        if (isset($options['t']))
            $table = $options['t'];
    }
    $i = 0;
    $parser = new LogParser();
    $data = new ApacheLogController($username, $password, $servername, $db, $table);

    echo "Parsing file: " . $log_file . "\n";
    $handle = fopen($log_file, "r");
    $done = 0;
    if ($handle) {
        while (($line = fgets($handle)) !== false) {
            $entry = $parser->parse($line);
            $data->save_log($entry);
            $done++;
        }
        fclose($handle);
    } else {
        echo "An error occurred: unable to open the file!\n";
    }
} catch (Exception $e) {
    echo 'An error occurred: ' . $e->getMessage() . "\n";
}

echo "Script executed\n";


$ru = getrusage();
echo $footer;
echo "Computations \t:" . $data->rutime($ru, $rustart, "utime") .
    " ms\n";
echo "System calls \t:" . $data->rutime($ru, $rustart, "stime") .
    " ms\n";
echo "Total lines \t:" . $done . "\n";
echo $footer2;



LogParser.php
<?php

/**
 * Created by PhpStorm.
 * User: kharr
 * Date: 22/02/2017
 * Time: 21:09
 */
class LogParser
{
    protected static $defaultFormat = '%h %l %u %t "%r" %>s %b';
    protected $pcreFormat;
    protected $patterns = array(
        '%%' => '(?P<percent>\%)',
        '%a' => '(?P<remoteIp>)',
        '%A' => '(?P<localIp>)',
        '%h' => '(?P<host>[a-zA-Z0-9\-\._:]+)',
        '%l' => '(?P<logname>(?:-|[\w-]+))',
        '%m' => '(?P<requestMethod>OPTIONS|GET|HEAD|POST|PUT|DELETE|TRACE|CONNECT|PATCH|PROPFIND)',
        '%p' => '(?P<port>\d+)',
        '%r' => '(?P<request>(?:(?:[A-Z]+) .+? HTTP/1.(?:0|1))|-|)',
        '%t' => '\[(?P<time>\d{2}/(?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)/\d{4}:\d{2}:\d{2}:\d{2} (?:-|\+)\d{4})\]',
        '%u' => '(?P<user>(?:-|[\w-]+))',
        '%U' => '(?P<URL>.+?)',
        '%v' => '(?P<serverName>([a-zA-Z0-9]+)([a-z0-9.-]*))',
        '%V' => '(?P<canonicalServerName>([a-zA-Z0-9]+)([a-z0-9.-]*))',
        '%>s' => '(?P<status>\d{3}|-)',
        '%b' => '(?P<responseBytes>(\d+|-))',
        '%T' => '(?P<requestTime>(\d+\.?\d*))',
        '%O' => '(?P<sentBytes>[0-9]+)',
        '%I' => '(?P<receivedBytes>[0-9]+)',
        '\%\{(?P<name>[a-zA-Z]+)(?P<name2>[-]?)(?P<name3>[a-zA-Z]+)\}i' => '(?P<Header\\1\\3>.*?)',
        '%D' => '(?P<timeServeRequest>[0-9]+)',
    );


    public static function getDefaultFormat()
    {
        return self::$defaultFormat;
    }

    public function __construct($format = null)
    {
        // Set IPv4 & IPv6 recognition patterns
        $ipPatterns = implode('|', array(
            'ipv4' => '(((25[0-5]|2[0-4][0-9]|[01]?[0-9]?[0-9])\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9]?[0-9]))',
            'ipv6full' => '([0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){7})', // 1:1:1:1:1:1:1:1
            'ipv6null' => '(::)',
            'ipv6leading' => '(:(:[0-9A-Fa-f]{1,4}){1,7})', // ::1:1:1:1:1:1:1
            'ipv6mid' => '(([0-9A-Fa-f]{1,4}:){1,6}(:[0-9A-Fa-f]{1,4}){1,6})', // 1:1:1::1:1:1
            'ipv6trailing' => '(([0-9A-Fa-f]{1,4}:){1,7}:)', // 1:1:1:1:1:1:1::
        ));
        $this->patterns['%a'] = '(?P<remoteIp>' . $ipPatterns . ')';
        $this->patterns['%A'] = '(?P<localIp>' . $ipPatterns . ')';
        $this->setFormat($format ?: self::getDefaultFormat());
    }

    public function parse($line)
    {
        if (!preg_match($this->pcreFormat, $line, $matches)) {
            echo "No match\n";
        }
        $entry = new \stdClass();
        foreach (array_filter(array_keys($matches), 'is_string') as $key) {
            if ('time' === $key && true !== $stamp = strtotime($matches[$key])) {
                $entry->stamp = $stamp;
            }
            $entry->{$key} = $matches[$key];
        }
        return $entry;
    }


    public function setFormat($format)
    {
        $expr = "#^{$format}$#";
        foreach ($this->patterns as $pattern => $replace) {
            $expr = preg_replace("/{$pattern}/", $replace, $expr);
        }
        $this->pcreFormat = $expr;
    }

}

Controller/ApacheLogController.php
<?php

class ApacheLogController
{
    public $config;
    public $conn;
    private $username;
    private $password;
    private $servername;
    private $db;
    private $table;

    function __construct($user, $pass, $server, $db, $table)
    {
        try {
            echo "Initialising Data storage...\n";
            $this->username = $user;
            $this->password = $pass;
            $this->servername = $server;
            $this->db = $db;
            $this->table = $table;
            $this->conn = new PDO("mysql:host=$this->servername", $this->username, $this->password,
                array(PDO::MYSQL_ATTR_INIT_COMMAND => "SET NAMES utf8"));
            $this->conn->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
            $sql = "CREATE DATABASE IF NOT EXISTS " . $this->db;
            $this->conn->exec($sql);
            $sql = "use " . $this->db;
            $this->conn->exec($sql);
            $sql = "CREATE TABLE IF NOT EXISTS `" . $this->db . "`.`" . $this->table . "` ( `id` INT(50) NOT NULL AUTO_INCREMENT , `host` VARCHAR(255) NOT NULL , `logname` VARCHAR(255) NOT NULL , `user` VARCHAR(255) NOT NULL , `stamp` VARCHAR(255) NOT NULL , `time` VARCHAR(255) NOT NULL , `request` TEXT NOT NULL , `status` VARCHAR(255) NOT NULL , `responseBytes` VARCHAR(255) NOT NULL , PRIMARY KEY (`id`)) ENGINE = MyISAM;";
            $this->conn->exec($sql);
            echo "Data storage ready...\n";
        } catch (PDOException $e) {
            echo 'ERROR: ' . $e->getMessage() . "\n";
        }
    }


    function save_log($data)
    {
        try {
            $data = (array)$data;
            $sql = "INSERT INTO `" . $this->db . "`.`" . $this->table . "` ( `host`, `logname`, `user`, `stamp`, `time`, `request`, `status`, `responseBytes`) VALUES ( :host, :logname, :user_id, :stamp, :time_id, :request, :status, :responseBytes)";
            $stmt = $this->conn->prepare($sql);
            $stmt->bindParam(':host', $data['host'], PDO::PARAM_STR);
            $stmt->bindParam(':logname', $data['logname'], PDO::PARAM_STR);
            $stmt->bindParam(':user_id', $data['user'], PDO::PARAM_STR);
            $stmt->bindParam(':stamp', $data['stamp'], PDO::PARAM_STR);
            $stmt->bindParam(':time_id', $data['time'], PDO::PARAM_STR);
            $stmt->bindParam(':request', $data['request'], PDO::PARAM_STR);
            $stmt->bindParam(':status', $data['status'], PDO::PARAM_STR);
            $stmt->bindParam(':responseBytes', $data['responseBytes'], PDO::PARAM_STR);
            $stmt->execute();

        } catch (PDOException $e) {
            echo "Error: " . $e->getMessage() . "\n";
        }
    }

    function rutime($ru, $rus, $index)
    {
        return ($ru["ru_$index.tv_sec"] * 1000 + intval($ru["ru_$index.tv_usec"] / 1000))
            - ($rus["ru_$index.tv_sec"] * 1000 + intval($rus["ru_$index.tv_usec"] / 1000));
    }
}

Conf/config.php
<?php

$username = 'root';
$password = '';
$servername = 'localhost';
$db = 'apache_log';
$table = 'log-' . time();
$log_file = 'data/access_log';
$header = "\n------------------------------- Access Log Analyser -----------------------------------\n";
$help = "
Name

    parser - PHP Command Line to parse the access log

Synopsis

    php parser.php [ -f ] file [ -d ] database [ -t ] table

Description

    This command line can be used to execute a parser which reads the Apache access log, parses it and saves it to a
    database
    It parses a standard apache access.log file as reference
    It saves the input into a Mysql Table
    LogFormat %h %l %u %t % r %>s %b common
    Log sample line
        127.0.0.1 - frank [10/Oct/2000:13:55:36 -0700] \"GET /apache_pb.gif HTTP/1.0\" 200 2326

Options

    -f:\t(optional) Contains the url to the file
    \tdefault value: /data/access_log
    -d:\t(optional) Contains the Database name (created if not exist)
    \tdefault value: apache_log
    -t:\t(optional) Contains the table name (created if not exist)
    \tdefault value: log-(timestamp)

Examples

php parser.php'
\tThis command simply Executes the script with default values

php parser.php -f \"path_to_file\" -d \"sampleDB\" -t \"sapmleTable\"
\tThis command parses file path_to_file and saves data inside sampleTable within sampleDB\n
-----------------------------------------------------------------------------------------\n";
$footer = "For more options Type : php parser.php -h\n
------------------------------------------ Stats -------------------------------------\n";
$footer2 = "--------------------------------------------------------------------------------------\n";

